### PR TITLE
ENT-10273: Upgrade BC to 1.75.

### PR DESCRIPTION
**Pull request by adelel1**
Upgrade BC to 1.75

### CODE REVIEW

File: core-deterministic/build.gradle [hunk 1]
```diff
@@ -49,8 +49,8 @@ dependencies {
 
     // These dependencies will become "runtime" scoped in our published POM.
     // See publish.dependenciesFrom.defaultScope.
-    deterministicLibraries "org.bouncycastle:bcprov-jdk15on:$bouncycastle_version"
-    deterministicLibraries "org.bouncycastle:bcpkix-jdk15on:$bouncycastle_version"
+    deterministicLibraries "org.bouncycastle:bcprov-jdk18on:$bouncycastle_version"
+    deterministicLibraries "org.bouncycastle:bcpkix-jdk18on:$bouncycastle_version"
```
Comment (chrisr3) [inline core-deterministic/build.gradle]: I'm assuming this means that these BC artifacts are for JDK 1.8, i.e. Java 8?

Comment (rick-r3) [reply core-deterministic/build.gradle]: Yes.   There is now a `15to18` for Java 1.5 to 1.8 and an `18on` which is 1.8+.   Slightly unfortunate we are now also into the realms of 15 and 18 as versions to add confusion!  But can confirm the BC build files say 1.5 and 1.8.

 Looks like decoding an RSA_SHA256 key now involves some non-determinism as we get an exception with the message "Seed generation disabled" from BC 1.71 onwards. This looks like a DJVM exception message.

**Comment by shamsasari** [comment]
The reason the DJVM serialisation tests are failing is due to extra checks in the construction of RSA keys and ECCurve objects (which are created when `BCECPublicKey.hashCode()` is called) where random numbers are used to check if certain parameters are prime (https://github.com/bcgit/bc-java/commit/9b5bc534ca9c40ce28c57b874f0d9f07b5c2fdd3).

**Comment by chrisr3** [comment]
> The reason the DJVM serialisation tests are failing is due to extra checks in the construction of RSA keys and ECCurve objects (which are created when `BCECPublicKey.hashCode()` is called) where random numbers are used to check if certain parameters are prime ([bcgit/bc-java@9b5bc53](https://github.com/bcgit/bc-java/commit/9b5bc534ca9c40ce28c57b874f0d9f07b5c2fdd3)).

Can you determine how BC discovers that "Seed generation [is] disabled"? Because, well, we're talking about the deterministic sandbox and so it probably _is_. And quite deliberately.

**Comment by shamsasari** [comment]
The error seems to be coming from the deterministic-rt.jar: https://github.com/corda/openjdk/blame/deterministic-jvm8/jdk/src/share/classes/sun/security/provider/SecureRandom.java#L136

**Comment by chrisr3** [comment]
> The error seems to be coming from the deterministic-rt.jar: https://github.com/corda/openjdk/blame/deterministic-jvm8/jdk/src/share/classes/sun/security/provider/SecureRandom.java#L136

Ah. _Alas, poor Yoric!_ :skull:

You'd probably better put this in `serialization-djvm/build.gradle` then:
```gradle
configurations {
    testRuntimeClasspath {
        resolutionStrategy {
            dependencySubstitution {
                substitute module('org.bouncycastle:bcprov-jdk18on') with module('org.bouncycastle:bcprov-jdk18on:1.70')
                substitute module('org.bouncycastle:bcpkix-jdk18on') with module('org.bouncycastle:bcpkix-jdk18on:1.70')
            }
        }
    }
}
```
Unless you _want_ to modify `deterministic-rt.jar` so that it creates a deterministic instance of `SecureRandom`?

**Comment by shamsasari** [comment]
The module substitution at test runtime doesn't work since the problematic classes are in the core-deterministic jar. I've reverted the BC dependency in core-deterministic jar back to 1.70, which is effectively the same thing.

**Comment by chrisr3** [comment]
> The module substitution at test runtime doesn't work since the problematic classes are in the core-deterministic jar. I've reverted the BC dependency in core-deterministic jar back to 1.70, which is effectively the same thing.

That doesn't make any sense! The _whole point_ of the substitution is to replace the version of BC used by the tests so that it doesn't invoke `SecureRandom` in the first place!

`core-deterministic` does not contain `deterministic-rt.jar`.

**Comment by shamsasari** [comment]
The invocation of `SecureRandom` is done inside the sandbox during deserialisation.

**Comment by chrisr3** [comment]
> The invocation of `SecureRandom` is done inside the sandbox during deserialisation.

Yes, but it is BC 1.75 that invokes `SecureRandom`, and so the only way this error can happen is if the sandbox is _still_ using BC 1.75.

Maybe dependency substitution isn't the way to do this. Perhaps:
```gradle
dependencies {
    constraints {
        testImplementation('org.bouncycastle:bcprov-jdk18on') {
            version {
                strictly '1.70'
            }
        }
        testImplementation('org.bouncycastle:bcpkix-jdk18on') {
            version {
                strictly '1.70'
            }
        }
    }
}
```
would be more effective?

**Comment by shamsasari** [comment]
I tried adding that block both to `serialization-djvm` and `core-deterministic` and neither worked, which isn't surprising as the BC dependencies are not in the test classpath.

Regardless, if we're willinging to downgrade the test to use BC 1.70 then we're only supporting the sandbox with BC 1.70 anyway. So might as well make that explicit.

**Review by rick-r3 — COMMENTED** [review]
All good.

**Comment by chrisr3** [comment]
> I tried adding that block both to `serialization-djvm` and `core-deterministic` and neither worked, which isn't surprising as the BC dependencies are not in the test classpath.

_Of course they're in the `testRuntimeClasspath`_! You can prove that just by iterating over its contents and printing them out. The problem is more likely that BC is entering the `testRuntimeClasspath` from multiple configurations and the "dependency substitution" would need to catch all of them before `testRuntimeClasspath` is resolved. Which is why I suggested trying to use a constraint on the resolve step instead.

Isn't Gradle "fun"?

**Pull request merged**